Transactions logic upgrade

symbolist opened the pull request:
https://openedx.atlassian.net/browse/TNL-3428

As part of the Django upgrade, we need to move to the transaction framework that was introduced in Django 1.7.

https://docs.djangoproject.com/en/1.8/topics/db/transactions/

This Stackoverflow answer explains the differences of atomicity and durability between the two APIs: http://stackoverflow.com/a/24108491

The different cases I came across and how they are being handled is as follows:

The most important constraint to keep in mind is that we cannot have commit() or rollback() calls inside an atomic block. This means if a function has either of these calls any view which can call it cannot be wrapped in atomic.

1. Most of the views did not have any custom transaction management. We do not need to do anything besides enabling the new API for wrapping the views.
2. Some of the functions which were calling commit() or rollback() can just use the atomic context manager. This does change the behaviour sometimes because previously a commit would happen when commit() was called but if the atomic context manager is nested inside another, the commit would happen at the end of the outer atomic. I think in some cases this a reasonable thing to do. student/view.py:confirm_email_change is one such case.
3. In 3 cases we are doing a commit() if an IntegrityError happens followed by a database read. In all these cases this was quite deep in the call stack and those particular functions can be called from multiple places. I cannot think of a good way to retain this behaviour and have removed it for now.
4. In grades.py we do commit calls frequently to lesson the chance of overlap between multiple concurrent requests and hence IntegrityErrors. Similarly in the instructor_task app, we commit after creating InstructorTask objects and before sending off the task to celery. In these cases, I think the right approach would be to not have these views wrapped in atomic and instead have a context manager which behaves similarly to the current commit_on_success() which can wrap the views and any functions in the call graph. The goal is that transaction managements in these views should continue to behave like it does now.
5. A number of other functions in the code base are decorated with commit_on_success(). But since these are called from multiple views, for now these have been wrapped in atomic(). However, this atomic is nested inside the atomic which wraps the calling view and so the commits would happen once the view returns instead of when these functions returns as it does now. The only alternative is use the solution for 4 in a lot of places and I have opted to avoid this if I think we can.

macdiesel commented:
@symbolist is this ready for review?

symbolist commented:
@macdiesel Not yet.

symbolist commented on common/djangoapps/student/management/commands/transfer_students.py:
```diff
@@ -52,7 +52,6 @@ class Command(TrackedCommand):
                     help="If True, try to transfer certificate items to the new course.")
     )
 
-    @transaction.commit_manually
```
No need for this.

doctoryes replied:
Nit: Maybe just change this to `@transaction.atomic` and remove the `commit_on_success` context mgr below?

symbolist replied:
Actually I have already replaced `commit_on_success` with `atomic` below. The reason for keeping it inside the loop is that change for each student should get committed as it happens. Otherwise an error would revert all of the work. Or do you think we should be doing that?

macdiesel replied:
@symbolist @doctoryes since the transaction isn't in a try block. Does that mean that if command fails on a user part way through the transaction the database would end up in an inconsistent state as there is nothing to catch the exception? You would end up with all the students to that point transferred but the remainder not transferred. 

Also would this end up failing on the rerun as the student has already been transferred to the new course?

We have two routes of action here:
1) add the try/catch and migrate students with no errors and report the ones that were not moved 
or
2) fail the entire transaction and keep all students in a consistent state.

I think that #2 is better as it would then allow us to troubleshoot and fix the issue without any inconsistencies in the data.

symbolist replied:
@macdiesel "if command fails on a user part way through the transaction" `atomic()` will 1. catch the exception, 2. rollback the transaction and 3. re-raise the exception. And yes, that means some of the students would have been transferred and the others not. But this is how it currently works too.

Actually, in Django 1.4 `commit_manually` is only used to disable automatic transaction management for views. Commands do not need it. So its not doing anything right now either. 
https://docs.djangoproject.com/en/1.4/topics/db/transactions/#django.db.transaction.commit_manually

doctoryes replied:
In this case, I think I lean towards 2) as well - all students transfer or no students transfer. However, that's different than the way the command currently works. And, unless it's an egregious error, we should likely stick with how it's always worked. So my final vote is for @symbolist 's change.

doctoryes replied:
BTW, @symbolist - have you seen @nedbat 's following commit?:

https://github.com/edx/edx-platform/commit/85c57f4fb54c7352f1e3b796be46ad01b0fce17d

It changes several `commit_on_success` to `atomic`.

symbolist replied:
@doctoryes I created a branch, reverted @nedbat's commit and then created this PR against it. So all the migration is visible in this PR. I thought it would give us a picture of all of the transaction related logic in one place.

macdiesel replied:
I do not know how this all plays out in the large scheme of how students and courses are handled in edx. I realize we would be changing behavior of this command, but I would rather see us keep the state of our data consistent. 

Let me pose this question what will be easier to track down in the future:
1) a command that fails to move all the students and provides a clear and concise error message
or
2) a command that succeeds, maybe puts something to stdout or stderr that someone will read or maybe not, then later when someone realizes that not all the students are moved they call in support or devops. Then someone has to comb through the data looking for said students, then they have to look to the code to try and figure out what happened. 

I just want to see us strive for both consistency in both our data and program behavior.

/soapbox

doctoryes replied:
Hey - let me on that soapbox! :smile: 

Totally agree that 1) is better code/procedure. Just worried about the "hey - this command changed - rollback the release!" scenario. If we're willing to take on some of that risk (maybe release notes would help?), let's do it.

symbolist replied:
I have added a `transaction.atomic` on the command itself. This is a rarely used command so it shouldn't be an issue.

symbolist replied:
See below.

symbolist commented on common/djangoapps/student/models.py:
```diff
@@ -909,16 +910,11 @@ def get_or_create_enrollment(cls, user, course_key):
-            enrollment = CourseEnrollment.objects.get(
-                user=user,
-                course_id=course_key,
-            )
```
This is called from too many views which have the `atomic` decorator and that does not allow `commits()` inside it. We will have to work out a different solution later if this is really a problem.

adampalay replied:
We can keep an eye on this in the logs—I think I remember this used to happen a good deal

symbolist replied:
Yes, the first solution we should try is to use `READ COMMITTED` for the views which cause IntegrityErrors here. Otherwise, we will have to do a good deal of refactoring.

symbolist commented on common/djangoapps/student/views.py:
```diff
@@ -2133,18 +2133,17 @@ def do_email_change_request(user, new_email, activation_key=None):
         except PendingEmailChange.DoesNotExist:
             response = render_to_response("invalid_email_key.html", {})
-            transaction.rollback()
+            set_rollback(True)
```
`set_rollback(True)` forces a rollback on exiting the `atomic` decorator.

symbolist commented on common/djangoapps/student/views.py:
```diff
@@ -2186,16 +2185,11 @@ def confirm_email_change(request, key):  # pylint: disable=unused-argument
-    except Exception:  # pylint: disable=broad-except
-        # If we get an unexpected exception, be sure to rollback the transaction
-        transaction.rollback()
-        raise
```
The `atomic` logic that wraps the view will take care of this.

doctoryes replied:
Agreed.

symbolist commented on lms/djangoapps/ccx/overrides.py:
```diff
@@ -143,15 +143,15 @@ def override_field_for_ccx(ccx, block, name, value):
+                )
+                _get_overrides_for_ccx(ccx).setdefault(block.location, {})[name + "_id"] = override.id
         except IntegrityError:
-            transaction.commit()
```
This function can be called from a number of views which are `atomic` so we can't do this here.

symbolist replied:
This is called from three views. We can add change the isolation level to `read_committed` for them.

symbolist commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -740,19 +736,6 @@ def get_score(user, problem_descriptor, module_creator, scores_client, submissio
-        raise
-    else:
-        transaction.commit()
-
```
The only difference between this and Django's `commit_on_success` was that this had logging. Lets just use the version of `commit_on_success` from `utils/db.py` instead.

symbolist commented on lms/djangoapps/instructor_task/subtasks.py:
```diff
@@ -526,65 +525,62 @@ def _update_subtask_status(entry_id, current_task_id, new_subtask_status):
         raise
-    else:
-        TASK_LOG.debug("about to commit....")
-        transaction.commit()
```
The `atomic` context manager takes care of the commit/rollback.

doctoryes commented on lms/djangoapps/instructor_task/subtasks.py:
```diff
@@ -494,7 +494,6 @@ def update_subtask_status(entry_id, current_task_id, new_subtask_status, retry_c
         _release_subtask_lock(current_task_id)
 
 
-@transaction.commit_manually
```
Why not just use the `@transaction.atomic` decorator here and avoid the reformatting below?

doctoryes replied:
BTW, this change from the deprecated `commit_manually` to `atomic` will fix a few bulk email `False is not true` failures.

symbolist replied:
Yes, I'll do that.

symbolist replied:
Updated below.

symbolist commented on common/djangoapps/util/testing.py:
```diff
@@ -90,3 +93,23 @@ def reset_tracker(self):
+        return classmethod(_w)
+
+    TestCase._enter_atomics = atomics_wrapper(TestCase._enter_atomics, False)
+    TestCase._rollback_atomics = atomics_wrapper(TestCase._rollback_atomics, True)
```
Inspired by https://github.com/django/django/blob/1.4.22/django/test/testcases.py#L61-L85.

symbolist commented on common/djangoapps/util/db.py:
```diff
@@ -4,44 +4,119 @@
 
 
-def commit_on_success_with_read_committed(func):  # pylint: disable=invalid-name
+class CommitOnSuccessManager(object):
```
This is mostly taken from https://github.com/django/django/blob/1.8.5/django/db/transaction.py#L110-L277 but instead of using save points we are using `commit()` and that simplifies a lot of things.

The goal is to be as close as possible to https://github.com/django/django/blob/1.4.22/django/db/transaction.py#L263-L289

doctoryes replied:
My understanding here: this manager does *not* use savepoints - and either commits all DB queries in a possibly-nested transaction -or- none of them. It differs from `transaction.atomic` in that nested `atomic` calls can cause part of the transaction's queries to still be committed.

symbolist replied:
Right, though I would add that in case of nested `CommitOnSuccessManager`s, it may also happen that only part of the outer block's queries get committed. I linked this in the email and explains the difference between the two very nicely: http://stackoverflow.com/questions/21861207/is-transaction-atomic-same-as-transaction-commit-on-success/24108491#24108491

doctoryes commented on common/djangoapps/course_modes/views.py:
```diff
@@ -135,6 +136,7 @@ def get(self, request, course_id, error=None):
 
         return render_to_response("course_modes/choose.html", context)
 
+    @method_decorator(transaction.non_atomic_requests)
```
I don't understand why both `commit_on_success_with_read_committed` and `transaction.non_atomic_requests`. Seems to have the effect of setting the transaction isolation only?

symbolist replied:
By default all views which are not decorated with `transaction.non_atomic_request` get wrapped in `transaction.atomic`, which is why we need to add this. See https://github.com/django/django/blob/1.8.5/django/core/handlers/base.py#L129-L132

Now that Django will not do any transaction management for this view, we wrap it `commit_on_success_with_read_committed` to setup our own custom transaction mangement.

Does that make sense?

doctoryes replied:
Makes sense.

doctoryes commented on common/djangoapps/course_modes/views.py:
```diff
@@ -135,6 +136,7 @@ def get(self, request, course_id, error=None):
         return render_to_response("course_modes/choose.html", context)
 
+    @method_decorator(transaction.non_atomic_requests)
     @method_decorator(login_required)
```
This decorator should not be between the two others - and it should likely be in the place where it's executed first before any transaction setup. That location would be closest to the `def post()`.

symbolist replied:
If we want it execute before any transaction setup, wouldn't we want it to be the outermost one? However, given the logic in https://github.com/django/django/blob/1.8.5/django/core/handlers/base.py#L129-L132, I think `transaction.non_atomic_request` must be at the outermost decorator.

doctoryes replied:
Right - I got application order and execution order mixed up. Order seems right.

symbolist commented on lms/djangoapps/instructor_task/api.py:
```diff
@@ -95,12 +99,10 @@ def submit_rescore_problem_for_student(request, usage_key, student):  # pylint:
+    When called from any function that is wrapped in a transaction, an
+    autocommit buried within here will cause any pending transaction to
+    be committed by a successful save here. Any future database operations
+    will take place in a separate transaction.
```
I'll add it to the documentation as well, but we need this because we need to commit the `InstructorTask` object before handing of the task to the queue. And that means a mid-request commit.

symbolist replied:
The commits happens in `instructor_task/models.py` and `bulk_email/models.py`.

doctoryes commented on common/djangoapps/util/db.py:
```diff
@@ -4,44 +4,119 @@
+        return decorated
+
+
+def commit_on_success(using=None, read_committed=False):
```
Although this decorator would enable the move from Django 1.4->1.8, its usage should be __heavily discouraged__ post-release. All coders should be educated about Django 1.8 transactional techniques instead - and should convert code using this decorator to the new transactional model. Do you agree? If so, can you add words to this effect in the docstring?

macdiesel replied:
I'm a +1 on this. I would like to see us discourage things like this in the platform and backlog a discovery item for us to come up with a plan to better deal with the situations that currently require us to do things such as this.

doctoryes commented:
PHEW! I've finished my first pass and left comments. Some points:
* I agree that the added commit_on_success decorator will mirror the current Django 1.4 behavior.
* I'd prefer to make all transactions conform to Django1.8-style savepoints and the like. But it's unclear how much work that would be - likely significant.
* We'll still be able to identify the places where we've done this - and fix them in the FUUUUTUUURE (yes - a future that may never exist) if we choose to do so.
* More testing, demonstrating expected behavior in different nesting situations, is needed.

symbolist commented on lms/djangoapps/ccx/overrides.py:
```diff
@@ -142,18 +142,15 @@ def override_field_for_ccx(ccx, block, name, value):
-            transaction.commit()
-            kwargs = {'ccx': ccx, 'location': block.location, 'field': name}
-            override = CcxFieldOverride.objects.get(**kwargs)
+        else:
```
Instead of trying to `create` an object first and do a `get` on an `IntegrityError`, lets just do a `get` first.

adampalay replied:
@symbolist , would this increase the likelihood of race conditions?

symbolist replied:
@adampalay Yes, to an extent. But this function is only called from 3 Instructor dashboard CCX tab views so I doubt we will be seeing many parallel requests causing `IntegrityErrors`. If we do we can refactor the views a bit so that we can put `outer_atomic` around this whole function or maybe shift to `READ COMMITTED` for the views which do create `IntegrityErrors`.

I have put in a ticket for this: https://openedx.atlassian.net/browse/TNL-3614

symbolist commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -137,7 +137,6 @@ def get_cohorted_commentables(course_key):
     return ans
 
 
-@transaction.commit_on_success
```
There are no set of queries that need atomicity so we don't really need to replace this with `atomic` and this function is called on every courseware page request, so to avoid an extra pair of savepoint setup teardown queries, its better to just remove it.

symbolist commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -1105,18 +1105,8 @@ def get_or_create_verification_checkpoint(cls, course_id, checkpoint_location):
-            transaction.commit()
-            return cls.objects.get(course_id=course_id, checkpoint_location=checkpoint_location)
+        checkpoint, __ = cls.objects.get_or_create(course_id=course_id, checkpoint_location=checkpoint_location)
+        return checkpoint
```
I have changed the view below to use `read_committed` so we shouldn't need this anymore.

symbolist commented:
@doctoryes @macdiesel I have addressed all comments. Besides that, I have:
1. Made tweaks in a number of places.
2. Reduced the usages of `commit_on_success` to just the progress page views in `courseware`.
3. Fixed and added tests for `commit_on_success` and `outer_atomic`.

This is ready for another review.

symbolist commented:
@adampalay Here is the PR I mentioned today. If you would like to discuss any of the changes, please ping me.

doctoryes commented:
Nit: There's no docstring or comments explaining why the `outer_atomic()` decorator exists *and* the pattern that developers are expected to follow for transactions in Django 1.8 post-upgrade. Such comments would be useful - perhaps they could be added to tests showing how outer_atomic works and demonstrating the recommended pattern.

doctoryes commented:
In other words, it should be clear what's deprecated.

doctoryes commented:
Also, do you plan to create a ticket for `commit_on_success` removal for the progress page methods in `lms/djangoapps/courseware/{grades|views}.py`? What blocks those from being removed?

doctoryes commented:
A few nits and follow-ons - but no blockers for merging: :+1:

symbolist commented on common/djangoapps/util/db.py:
```diff
@@ -4,44 +4,213 @@
+
+    Arguments:
+        using (str): the name of the database.
+        read_committed (bool): Whether to use read committed isolation level.
```
@doctoryes I have added more detailed documentation to `outer_atomic()`.

symbolist commented:
@doctoryes I have added the tests you suggested as well. 

Yes, I have created a ticket for replacing the remaining instances of `commit_on_success` with `atomic`. I haven't looked at the grade functions in courseware very closely yet. I don't see why it shouldn't be possible though we may need to refactor things a bit. But I would like to wait until we can get a sandbox hooked up to prod data before attempting that so that we can test with real data that everything works fine after the migration.

adampalay commented on common/djangoapps/util/db.py:
```diff
@@ -4,44 +4,213 @@
+                    try:
+                        connection.rollback()
+                    except Error:
+                        # An error during rollback means that something
```
will we want to know if the rollback didn't work?

symbolist replied:
I'll put in some logging statements here. Though ideally we want to get rid of this `commit_on_success` decorator before we deploy the upgrade. Right now it is only being used in `courseware/grades.py` and on the progress page view. I have created a separate story for it.

symbolist commented:
@doctoryes @adampalay  I have also eliminated the rest of the instances of `commit_on_success` in `lms/djangoapps/courseware/{grades|views}.py` and refactored the code to use `outer_atomic`. You can see the changes in the last two commits.

A bit of context: The reason for adding `commit_on_success` to the progress page view and the helper functions for grading was that because calculating the grade for a course can take a long time – 10-15 seconds is normal for large courses – student's often click the progress page link multiple times because it does not look like it is responding and this creates a strong chance of two parallel requests which try to do `get_or_create` as they try to initialize xblocks and save data for them. Adding `commit_on_success` in these functions divided up the work for calculating the grade into multiple smaller transactions which had a much lower chance of collision. When migrating to the new API I have kept this in mind and used `outer_atomic` to divide this work into a number of smaller transaction.

symbolist commented:
@doctoryes Can you review the changes in the last two commits?

macdiesel commented:
:+1: :sunglasses:

symbolist commented:
Closing and creating a new PR against the upgrade branch.
